samples/rust/src/main.rs

```diff
@@ -3,18 +3,62 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use std::env;
+use std::fs::File;
+use std::io::{Read, Write};
 use std::path::PathBuf;
 use std::thread;
 use std::time::Duration;
 
 use tempfile::Builder;
 
-use glean::{ClientInfoMetrics, Configuration};
+use flate2::read::GzDecoder;
+use glean::{net, ClientInfoMetrics, ConfigurationBuilder};
 
 pub mod glean_metrics {
     include!(concat!(env!("OUT_DIR"), "/glean_metrics.rs"));
 }
 
+#[derive(Debug)]
+struct MovingUploader(String);
+
+impl net::PingUploader for MovingUploader {
+    fn upload(
+        &self,
+        url: String,
+        body: Vec<u8>,
+        headers: Vec<(String, String)>,
+    ) -> net::UploadResult {
+        let mut gzip_decoder = GzDecoder::new(&body[..]);
+        let mut s = String::with_capacity(body.len());
+
+        let data = gzip_decoder
+            .read_to_string(&mut s)
+            .ok()
+            .map(|_| &s[..])
+            .or_else(|| std::str::from_utf8(&body).ok())
+            .unwrap();
+
+        let mut out_path = PathBuf::from(&self.0);
+        out_path.push("sent_pings");
+        std::fs::create_dir_all(&out_path).unwrap();
+
+        let docid = url.rsplit('/').next().unwrap();
+        out_path.push(format!("{docid}.json"));
+        let mut fp = File::create(out_path).unwrap();
+
+        // pseudo-JSON, let's hope this works.
```

Comment:
Confidence!

Reply:
what is json if not just some combination of `{`, `}` and `,`